File: web/src-tauri/types/src/pdf.rs
use std::collections::HashMap;

use serde::{Deserialize, Serialize};
use ts_rs::TS;

use crate::FormatType;

#[derive(Serialize, Deserialize, Debug, Clone, TS)]
#[ts(export, export_to = "pdf.ts")]
pub struct IBindingsPDFSection {
    pub id: f64,
    pub size: f64,
}

#[derive(Serialize, Deserialize, Debug, Clone, TS)]
#[ts(export, export_to = "pdf.ts")]
pub struct IBindingsPDFToc {
    pub label: String,
    pub href: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[ts(optional)]
    pub subitems: Option<Vec<IBindingsPDFToc>>,
}

#[derive(Serialize, Deserialize, Debug, Clone, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export, export_to = "pdf.ts")]
pub struct IBindingsPDFMetadata {
    pub title: String,
    pub author: String,
    pub cover: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[ts(optional)]
    pub language: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[ts(optional)]
    pub publisher: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[ts(optional)]
    pub contributor: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[ts(optional)]
    pub published: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[ts(optional)]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[ts(optional)]
    pub identifier: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[ts(optional)]
    pub modified: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[ts(optional)]
    pub rights: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[ts(optional)]
    pub subject: Option<String>,
}

#[derive(Serialize, Deserialize, Debug, Clone, TS)]
#[ts(export, export_to = "pdf.ts")]
pub struct IBindingsPDFBook {
    pub metadata: IBindingsPDFMetadata,
    #[serde(rename = "percentageProgress")]
    #[ts(rename = "percentageProgress")]
    pub percentage_progress: String,
    pub progress: HashMap<String, String>,
    pub format: FormatType,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[ts(optional)]
    pub toc: Option<Vec<IBindingsPDFToc>>,
    pub sections: Vec<IBindingsPDFSection>,
    pub id: String,
}

#[derive(Serialize, Deserialize, Debug, Clone, TS)]
#[ts(export, export_to = "pdf.ts")]
pub struct IBindingsPDFBookStructure {
    pub sections: Vec<IBindingsPDFSection>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[ts(optional)]
    pub toc: Option<Vec<IBindingsPDFToc>>,
    pub format: FormatType,
}

File: web/src/interfaces/book/pdf.ts
import type { FormatType } from '@bindings/format'

export interface IPDFMetadata {
  author?: string
  contributor?: string
  description: string
  identifier?: string
  language?: string
  publisher?: string
  rights?: string
  source?: string
  modfied?: string
  subject?: string
  title?: string
}
export type IPDFToc = { label: string; href: string; subitems?: Array<IPDFToc> }

export interface IZoomEvent {
  doc: Document
  scale: number
}

interface IPDFRenderedPage {
  src: string
  onZoom: (event: IZoomEvent) => Promise<void>
}

export interface IPDFSections {
  id: number
  load: () => Promise<IPDFRenderedPage>
  size: number
}
export type IToc = { label: string; href: string; subitems?: Array<IToc> }

export interface IPDFBookFile {
  format: FormatType
  metadata: IPDFMetadata
  toc?: IToc[]
  percentageProgress: string
  progress: Record<string, string>
  sections: any[]
  id: string
}

export interface IPDFBook extends IPDFBookFile {
  format: 'PDF'
  getCover(): Promise<Blob | null>
}
